portality/dao.py

```diff
@@ -862,37 +862,85 @@ def iterate_unstable(cls, q: dict = None, page_size: int = 1000, limit: int = No
         # TODO: Another candidate for swapping to iterate_pit (or rename iterate_pit to iterate when we're happy)
         return cls.iterate(MatchAllQuery().query(), page_size, limit, **kwargs)
 
     @classmethod
-    def iterall_unstable(cls, page_size=1000, stripe_field="id", striped=False, prefix_generator=None, prefix_size=4, limit=None, logger=None, **kwargs):
+    def iterall_unstable(cls, page_size=1000,
+                         stripe_field="id",
+                         striped=False,
+                         prefix_generator=None,
+                         prefix_size=4,
+                         limit=None,
+                         logger=None,
+                         must=None,
+                         **kwargs):
         def hex_prefixes(n=4):
             """ Generate a list of hex prefixes of length n """
             return [str(hex(i))[2:].zfill(n) for i in range(0, 16 ** n)]
 
         if striped:
-            count = 0
-            prefixes = prefix_generator(prefix_size) if prefix_generator is not None else hex_prefixes(prefix_size)
-            for prefix in prefixes:
+            q = None
+            if must is not None:
                 q = {
                     "query": {
-                        "prefix": {stripe_field: prefix}
+                        "bool": {
+                            "must": must
+                        }
                     }
                 }
-                if logger: logger(f"Exporting prefix: {prefix}")
+            total = cls.count(q)
+            if total == 0:
+                return
+
+            count = 0
+            prefixes = prefix_generator(prefix_size) if prefix_generator is not None else hex_prefixes(prefix_size)
+            empty_prefixes = []
+            for prefix in prefixes:
+                if must is None:
+                    q = {
+                        "query": {
+                            "prefix": {stripe_field: prefix}
+                        }
+                    }
+                else:
+                    q = {
+                        "query": {
+                            "bool": {
+                                "must": must + [
+                                    {
+                                        "prefix": {stripe_field: prefix}
+                                    }
+                                ]
+                            }
+                        }
+                    }
+
+                first = True
                 for record in cls.iterate_unstable(q, page_size, limit=limit, logger=logger, **kwargs):
                     count += 1
+                    if logger and first:
+                        if len(empty_prefixes) > 0:
+                            logger(f"Skipped empty prefixes: {empty_prefixes}")
+                            empty_prefixes = []
+                        logger(f"Exporting prefix: {prefix}")
+                        first = False
```

Comment:
Do we only ever flip `first` if there's a `logger` set?

Reply:
good spot!